### PR TITLE
Refactor lib/creator and integrate validation in migrate

### PR DESCRIPTION
**Pull request by ev1stensberg**
**Did you add tests for your changes?**
N/A

**If relevant, did you update the documentation?**

N/A

**Summary**

Lays the groundwork for a first release. Integrated Validation to
migrate, refactored creator for the init feature to not be circular and
made a cleaner structure to the project inside lib/creator.

**Does this PR introduce a breaking change?**
No

**Other information**

### CODE REVIEW

**Comment by ev1stensberg** [comment]
@okonet Better to open this in PR in a text editor to see the changes. Could you review once you find the time?

File: lib/creator/transformations/init-transforms.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+const j = require('jscodeshift');
+
+module.exports = function (promptOptions) {
+	const promptParams = Object.keys(promptOptions);
+	if (promptParams && promptParams.length) {
+		let ast = j('{}');
```
Comment (okonet) [inline lib/creator/transformations/init-transforms.js]: That's not the `Object` as you might expect. This is a block statement. See https://astexplorer.net/#/gist/8ae05b2650bc27d2c4923ef8e6f36241/0a32afa94ee7cd5ecd742958c815706d34201345

File: lib/creator/transformations/init-transforms.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+				});
+			return ast;
+		},
+		'output': function (j, ast) {
+			const value = options['output'];
+			ast.find(j.ObjectExpression)
+				.forEach(
+					p => p.value.properties.push(
+						j.property('init', j.literal('output'),
+						j.objectExpression([j.property('init', j.literal('filename'), j.literal(value))])))
```
Comment (okonet) [inline lib/creator/transformations/init-transforms.js]: You should check out the utils we've written for migrate feature: https://github.com/webpack/webpack-cli/blob/master/lib/transformations/utils.js#L82

File: lib/migrate.js [hunk 3]
```diff
@@ -65,9 +68,17 @@ module.exports = function transformFile(currentConfigPath, outputConfigPath, opt
 				])
 				.then(answers => {
 					if (answers['confirmMigration']) {
-						// TODO validate the config
-						fs.writeFileSync(outputConfigPath, result, 'utf8');
-						console.log(chalk.green(`✔︎ New webpack v2 config file is at ${outputConfigPath}`));
+						fs.writeFile(outputConfigPath, result, 'utf8', (err) => {
+							const webpackOptionsValidationErrors = validateSchema(webpackOptionsSchema, require(outputConfigPath));
```
Comment (okonet) [inline lib/migrate.js]: 👍

Comment (okonet) [inline lib/migrate.js]: Should this go to a separate PR?

Comment (ev1stensberg) [inline lib/migrate.js]: Probably not needed. Fixing the transform issues and let's merge and take it from there.

**Comment by ev1stensberg** [comment]
@pksjce I've started the transformations, but let's move it to another seperate branch after this gets merged. I'll open a wip for you to follow along.

File: lib/creator/index.js [hunk 4]
```diff
@@ -1,42 +1,36 @@
...
-			throw new WebpackOptionsValidationError(webpackOptionsValidationErrors);
-		} else {
-			process.stdout.write('\n' + chalk.green('Congratulations! Your new webpack config file is created!') + '\n');
-		}
-		*/
+module.exports = function creator(options) {
+	const env = yeoman.createEnv(null, null, new WebpackAdapter());
+
+	if(options) {
+		env.register(require.resolve(options), 'npm:app');
```
Comment (eliperelman) [inline lib/creator/index.js]: I'd probably rename `npm:app` to avoid any potential collisions by other packages that may have registered a Yeoman environment. Maybe `webpack:cli`?

File: lib/creator/index.js [hunk 5]
```diff
@@ -1,42 +1,36 @@
...
+
+		env.run('npm:app')
+		.on('end', () => {
```
Comment (eliperelman) [inline lib/creator/index.js]: This is an arrow function, so you could either just one-line it and leave out the return, or since it's an event, you probably don't need the return at all.

Comment (ev1stensberg) [inline lib/creator/index.js]: This is actually just the first integration, would need to change the `generatorName` through `env.getGeneratorMetaData()` later, could integrate it now though.

File: lib/creator/index.js (hunk 4 above)
Comment (ev1stensberg) [inline lib/creator/index.js]: It's gonna be assigned to the generator name :) See previous comment.

File: lib/migrate.js (hunk 3 above)
Comment (okonet) [inline lib/migrate.js]: I'm working on a fix PR and desperately need this! :)

File: lib/creator/index.js [hunk 6]
```diff
@@ -1,42 +1,51 @@
...
+*
+* Replaces the webpack-addons pattern with the end of the addons name merged
+* with 'generator'
+*
+* @param { String } name - name of the generator
+* @returns { String } name - replaced pattern of the name
+*/
+
+function replaceGeneratorName(name) {
+	return name.replace(
```
Comment (okonet) [inline lib/creator/index.js]: Mind adding tests for this one?

File: lib/creator/index.js [hunk 7]
```diff
@@ -1,42 +1,51 @@
...
-		}
-		*/
+module.exports = function creator(options) {
+	const env = yeoman.createEnv(null, null, new WebpackAdapter());
+	if(options) {
+		const generatorName = replaceGeneratorName(path.basename(options));
+		env.register(require.resolve(options), generatorName);
+
+		env.run(generatorName)
+		.on('end', () => {
```
Comment (okonet) [inline lib/creator/index.js]: Can the duplication be removed from these 2 code branches?

Also indentation is odd but this can be fixed as #102 is merged,

**Comment by pksjce** [comment]
Cool. I will work on a separate PR for transformations. My previous PR was meant for feedback. I was really not sure what approach would work with yeoman.
Also could you solve these 
[https://www.codacy.com/app/okonet/webpack-cli/pullRequest?prid=594576](https://www.codacy.com/app/okonet/webpack-cli/pullRequest?prid=594576)
Would be good to comment them out for now.

File: lib/creator/index.js (hunk 7 above)
Comment (ev1stensberg) [inline lib/creator/index.js]: Yeah, I was thinking about it actually, will do that.

File: lib/creator/index.js (hunk 6 above)
Comment (ev1stensberg) [inline lib/creator/index.js]: Sure, giveme 1h :) At school

**Comment by ev1stensberg** [comment]
There we go @okonet @pksjce . I've started the work on the AST's @pksjce , lemme open a wip with my stuff for us to work with.

File: lib/creator/index.test.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+function replaceGeneratorName(name) {
```
Comment (okonet) [inline lib/creator/index.test.js]: Should this be the real implementation? Export it and import here to remove code duplication, please.

File: lib/creator/index.test.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+	return name.replace(
+      /(webpack-addons)?([^:]+)(:.*)?/g, 'generator$2');
+}
+
+describe('replaceGeneratorName', () => {
+
+	it('should replace a pattern of an addon', () => {
+		const generatorName = replaceGeneratorName('webpack-addons-thefox');
+		expect(generatorName).toEqual('generator-thefox');
```
Comment (okonet) [inline lib/creator/index.test.js]: The regex looks more complex as just and covers more edge cases. Should we test against them?

Comment (ev1stensberg) [inline lib/creator/index.test.js]: Just an initial commit, can add more tests later and label an issue with `good first contribution`
